Preserve dimension order across indexes during ingestion

jon-wei opened the pull request:
Partially addresses issue #658 

Related to https://github.com/druid-io/druid-api/pull/68

This changes the RealtimePlumber and IndexGeneratorJob so that when an index is persisted, the next index created will inherit the dimension order built up by the previous index.

This is to better support arbitrary dimension orders when schema-less ingestion is used.

When merging indexes, the index merger will now attempt to find the largest common dimension ordering across the indexes, e.g.:

Index 1 dims: DimA, DimB
Index 2 dims: DimC
Index 3 dims: DimC, DimA, DimB

The ordering that will be used will be DimC, DimA, DimB as that encompasses all of the shorter orderings.

If no common order is found, the merger will fall back to the original lexicographic dimension ordering.

To support this change, the persisted indexes now include null dimensions and the dimension encoding dictionaries will now always have an entry for the empty string. This is done so that previously seen dimensions are not dropped in case an index does not receive any events with values for that dimension.

LIMITATIONS:
- Dimension order can be passed between the FireHydrants within a Sink, but not across Sinks.
- Dimension orders can be passed between indexes within a single partition during batch ingestion, but not across partitions.

nishantmonu51 commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -208,10 +209,18 @@ private static IncrementalIndex makeIncrementalIndex(
         tuningConfig.getRowFlushBoundary()
     );
+
+    if(oldIndex != null) {
```
loadDimensionOrder looks a bit hacky way to do the initialization of dimensionsOrder
can we pass in the schema with the correct dimensionsOrder ? , 
e.g newIndex = new OnheapIncrementalIndex(schema.withDimensionsSpec(dimensionsSpec with custom order from oldIndex),... )

jon-wei replied:
@nishantmonu51 

I considered implementing it that way initially, but it would require a change to druid-api.

The presence of a dimensions list within the DimensionsSpec currently determines the result of hasCustomDimensions(), which controls if schema-less ingestion is used. 

Perhaps there should be a later PR that changes the DimensionsSpec semantics so that the user can specify a dimensions list but still have schema-less dimension discovery enabled. Then the index initialization could be changed to use DimensionsSpec to initialize the dimensions order; this would also give the user some control over the dimension ordering during schema-less ingestion.

Thoughts?

nishantmonu51 replied:
I think it would be nice to have ability to specify ordering for a subset of dimensions which are known prior to ingestion and rest of the dimensions can be discovered during ingestion,
can you submit a github issue to track this and add a comment to the issue here in code indicating that it can be cleaned up once the issue is resolved ?

jon-wei replied:
@nishantmonu51 will do

jon-wei replied:
https://github.com/druid-io/druid/issues/2011

himanshug commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -271,19 +273,52 @@ public File merge(
+      Iterator<String> candidateIter = orderingCandidate.iterator();
+      for (String matchDim : index.getDimensionNames()) {
+        boolean matched = false;
+        while (candidateIter.hasNext()) {
```
does it make sense to create a Set representation of orderingCandidate so that this while loop can be avoided

jon-wei replied:
@himanshug What advantage does the Set provide over the while loop?

himanshug replied:
lookup in [hash]set would be O(1) while with while loop it is O(n)

jon-wei replied:
@himanshug I don't think a Set would be helpful here, since the ordering of dimensions matters, it's not enough to check for presence in the candidate.

It could be useful though to build a map of dimension->list index for the dimensions in the candidate, and the per-index check could then iterate through the dimensions present in each index, checking that those dimensions exist in the candidate and their list indices always increase through the iterations

It would reduce the number of dimension list iterations when some dimension lists are significantly smaller than the candidate list, but not so much when the lists are similarly sized.

himanshug replied:
i see, if orderingCandidate is [A,B,C,D]
and say index.getDimensionNames is [B,D] this will pass but if it was [D,B] then it wouldn't .
ok, but still wondering why  [A,B,C,D] would be any less optimal than lexicographic when index had [D,B] dimension ordering.

jon-wei replied:
@himanshug [A,B,C,D] ordering would be the same as lexicographic ordering

If there is no ordering that covers all sub orderings, then we have to generate an ordering that includes all dimensions, and lexicographic seemed like a sensible fallback choice.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -565,6 +566,32 @@ public Integer getDimensionIndex(String dimension)
+    int i = 0;
+    for (String dim : oldDimensionOrder) {
+      dimensionOrder.put(dim, i);
+      dimensions.add(dim);
```
do we need to check here id the dimension is already present in dimValues ?

jon-wei replied:
@nishantmonu51 I think it's necessary, dimValues will throw an exception if you add something that already exists, and dimensions is an ArrayList (don't want duplicates)

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -565,6 +566,32 @@ public Integer getDimensionIndex(String dimension)
+    return dimensionOrder;
+  }
+
+  public void loadDimensionOrder(LinkedHashMap<String, Integer> oldDimensionOrder)
```
can this method be called at any point or only before ingesting any row ? 
If this needs to be called before ingesting any row, can we add a sanity check to ensure its called before add is called ? maybe check size of index = 0 here.

jon-wei replied:
@nishantmonu51 I intended for it to be callable at any point; I just updated this code so it acquires a lock on dimensionOrdering before adding anything

nishantmonu51 commented on processing/src/test/java/io/druid/query/metadata/SegmentMetadataQueryTest.java:
```diff
@@ -100,10 +100,10 @@ public SegmentMetadataQueryTest()
                 null
             )
-        ), 71982
+        ), 73191
```
can you explain why this change ?

jon-wei replied:
@nishantmonu51 Null dimensions are now stored in the index on disk if the dimension was present in the dimension ordering, the sample has an empty dim called null_column, so the size returned by segment metadata query is larger now

himanshug commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -624,6 +624,16 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
             }
+
+            if (rowValue.size() == 1) {
+              if(rowValue.get(0) == 0) {
```
if (rowValue.size() == 1 && rowValue.get(0) == ) {
..
} ?

jon-wei replied:
@himanshug merged the two if statements

himanshug commented on server/src/test/java/io/druid/segment/realtime/plumber/RealtimePlumberSchoolTest.java:
```diff
@@ -393,26 +403,154 @@ public void run()
+    );
+
+    plumber.persist(committer);
+    while (!committed.get()) {
```
using a latch for committed might avoid the while loop

jon-wei replied:
@himanshug changed while loop to a latch

himanshug commented:
besides minor comments, looks good to me overall. however, this might increase the size of segments for some users.

himanshug commented on processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java:
```diff
@@ -385,6 +385,9 @@ public IndexedInts seek(String value)
         if (currVal == null) {
+          if(currIndex == dimSet.size()) {
+            return new EmptyIndexedInts();
+          }
```
why is this change needed?

jon-wei replied:
It handles the case where the only entry in dimSet is the null string; the seek needs to stop and return if the dimSet has been exhausted. The index == size check is there to distinguish between the initial state of the seeker and when it has retrieved the null string's entry (in both cases currVal is null)

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -520,7 +530,7 @@ protected void reduce(
         long startTime = System.currentTimeMillis();
 
-        Set<String> allDimensionNames = Sets.newHashSet();
+        Set<String> allDimensionNames = Sets.newLinkedHashSet();
```
This set gets populated by `inputRow.getDimensions()` which depends on the order in which rows appear. Are we sue the order gets preserved here? If I ask for `[a,b,c]` in my indexing spec, and if my first row only has dimension `c`, my second row only has dimension `b`, and my third row only has dimension `a`, then this set may end up with `[c,b,a]` instead of `[a,b,c]`

jon-wei replied:
@xvrl In the case where the user has specified a list of dimensions in the indexing spec, the MapInputRowParser initializes all of the MapBasedInputRows that it creates with the specified dimension list, so inputRow.getDimensions() would return the same order every time.

xvrl replied:
@jon-wei does it always return the same dimensions for every single row, and is that also the case for other InputRowParsers.

jon-wei replied:
@xvrl yes, every row returns the same dimensions if the user specifies the dimension list in the ingestion spec. There's only StringInputRowParser and MapInputRowParser which are coupled together.

xvrl replied:
in that case, why do we need to add the dimensions to the set for every single row?

jon-wei replied:
@xvrl for the schemaless case where dimensions are discovered on the fly

xvrl replied:
@jon-wei do we have a test for hadoop indexing that covers both:
- schema-less indexing to make sure that dimension are persisted in the order they were seen
- schema-full indexing where we ensuer the order in which dimensions appear when read is different than the ones specified in the spec, and we test that the persisted order corresponds to the spec ?

jon-wei replied:
@xvrl The test I added to IndexGeneratorJob covers the first case, I'll add another test that covers the second

jon-wei replied:
@xvrl Added a test to IndexGeneratorJobTest that checks original order is maintained for schema-full indexing

gianm commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -578,6 +579,43 @@ public Integer getDimensionIndex(String dimension)
 
+  public LinkedHashMap<String, Integer> getDimensionOrder()
+  {
+    return dimensionOrder;
```
synchronize on dimensionOrder, return a copy? (dimensionOrder is not a thread safe map)

gianm replied:
Actually… we don't need `dimensionOrder` at all for this, we can just return `dimensions` itself. So this can return a `List<String>` that is actually just `ImmutableList.copyOf(dimensions)`.

That also makes the loading simpler since you can pass that list directly to the one that takes an Iterable.

gianm commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -578,6 +579,43 @@ public Integer getDimensionIndex(String dimension)
+  public void loadDimensionOrder(LinkedHashMap<String, Integer> oldDimensionOrder)
+  {
+    synchronized (dimensionOrder) {
+      dimensionOrder.putAll(oldDimensionOrder);
```
throw ISE if dimensionOrder is not empty

jon-wei replied:
@gianm added ISE on non-empty dimensionOrder

gianm commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -578,6 +579,43 @@ public Integer getDimensionIndex(String dimension)
+  public void loadDimensionIterable(Iterable<String> oldDimensionOrder)
+  {
+    int i = 0;
+    synchronized (dimensionOrder) {
```
throw ISE if dimensionOrder is not empty

gianm commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -578,6 +579,43 @@ public Integer getDimensionIndex(String dimension)
+   * Index dimension ordering could be changed to initalize from DimensionsSpec after resolution of
+   * https://github.com/druid-io/druid/issues/2011
+   */
+  public void loadDimensionOrder(LinkedHashMap<String, Integer> oldDimensionOrder)
```
This method and `loadDimensionIterable` are doing basically much the same thing. How about removing this and just always using loadDimensionIterable?

Like, `loadDimensionIterable(aDimensionOrder.keySet())`

gianm replied:
Or, just switch `getDimensionOrder` to return a `List`

jon-wei replied:
@gianm changed getDimensionOrder to return an immutable list copy of 'dimensions'

jon-wei replied:
@gianm got rid of loadDimensionOrder, keeping loadDimensionIterable

gianm commented on server/src/main/java/io/druid/segment/realtime/plumber/Sink.java:
```diff
@@ -204,6 +207,15 @@ private FireHydrant makeNewCurrIndex(long minTimestamp, DataSchema schema)
       if (numHydrants > 0) {
         FireHydrant lastHydrant = hydrants.get(numHydrants - 1);
         newCount = lastHydrant.getCount() + 1;
+        if(!indexSchema.getDimensionsSpec().hasCustomDimensions()) {
```
apply code style

gianm commented on server/src/main/java/io/druid/segment/realtime/plumber/Sink.java:
```diff
@@ -204,6 +207,15 @@ private FireHydrant makeNewCurrIndex(long minTimestamp, DataSchema schema)
         FireHydrant lastHydrant = hydrants.get(numHydrants - 1);
         newCount = lastHydrant.getCount() + 1;
+        if(!indexSchema.getDimensionsSpec().hasCustomDimensions()) {
+          if (lastHydrant.hasSwapped()) {
```
I think the `lastHydrant` should never have swapped at any point that this method is being called.

jon-wei replied:
@gianm I am seeing that case get hit, can provide more details later

jon-wei replied:
@gianm lastHydrant.hasSwapped() is true when makeNewCurrentIndex is called from the bootstrapSinksFromDisk() path, when the hydrants are created using QueryableIndexSegment

gianm commented:
I think it should be possible to do this without writing out the null columns. To avoid that we need to be sure that:

- Whenever we merge indexes, the dimension orders of the individual indexes (some of which might be missing some dimensions) should be merged such that if any dimension A appears before a dimension B in all indexes, it must also appear before B in the final order.
- Whenever we create a new index, it's provided with an order that is based on all prior indexes, not just the immediately previous one.

jon-wei commented:
@gianm 

I've revised the PR to avoid persisting null columns. 

IndexGeneratorJob and Sink have been changed to remember dimension orders outside of/across their individual indexes.

I also added a change to IndexIO's v8 to v9 conversion function, I noticed in some testing/debugging that filtering on a null value would fail in some cases if I converted null dim values to empty strings before persisting an IncrementalIndex. A check for empty string values was added so the "dimensions with null values" logic would be handled properly.

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -210,10 +213,18 @@ private static IncrementalIndex makeIncrementalIndex(
         tuningConfig.getRowFlushBoundary()
     );
+
+    if(oldDimOrder != null) {
```
formatting issues here

fjy commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -210,10 +213,18 @@ private static IncrementalIndex makeIncrementalIndex(
     );
+
+    if(oldDimOrder != null) {
+      if (!indexSchema.getDimensionsSpec().hasCustomDimensions()) {
```
we can probably just use 1 if statement

jon-wei replied:
@fjy merged the if statements and fixed formatting

jon-wei commented:
@himanshug @gianm can we merge this

fjy commented:
:+1: had some minor comments but looks good otherwise

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -271,19 +274,52 @@ public File merge(
+    int maxSize = 0;
+    Iterable<String> orderingCandidate = null;
+    for (IndexableAdapter index : indexes) {
+      int iterSize = Iterators.size(index.getDimensionNames().iterator());
```
This could be `index.getDimensionNames().size()`

jon-wei replied:
@gianm changed

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -271,19 +274,52 @@ public File merge(
-      }
+    List<String> commonDimOrder = getLongestSharedDimOrder(indexes);
+    if (commonDimOrder == null) {
+      return getLexicographicMergedDimensions(indexes);
```
Perhaps do a `log.warn` here, as this case is not expected during normal indexing.

jon-wei replied:
@gianm added a warning log

gianm commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -618,17 +619,25 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
+                nullsSet = bitmapFactory.makeEmptyMutableBitmap();
+              }
+              nullsSet.add(i);
+              continue;
```
this `continue` seems unnecessary, is that right?

jon-wei replied:
@gianm Removed unneeded changes in IndexIO

gianm commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -618,17 +619,25 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
                 nullsSet = bitmapFactory.makeEmptyMutableBitmap();
               }
               nullsSet.add(i);
+              continue;
```
I think this `continue` / `if` could be an `else if`

gianm commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -618,17 +619,25 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
+              break;
             }
+
             if (rowValue.size() == 0) {
```
Could this be `rowValue.size() == 0 || rowValue.get(0) == emptyStrIdx)` instead of two if blocks?

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -555,19 +591,32 @@ private File makeIndexFiles(
       DimValueConverter[] converters = new DimValueConverter[indexes.size()];
+      Indexed<String> nullStrValues = null;
+
+      if(indexes.size() > 1) {
```
formatting

jon-wei replied:
@gianm fixed formatting

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -555,19 +591,32 @@ private File makeIndexFiles(
       dimConversions.add(Maps.<String, IntBuffer>newHashMap());
     }
 
+    Map<String, Integer> nullStringIndexMap = new LinkedHashMap<>();
```
It doesn't look like this needs to be Linked, if not please just use a regular HashMap

jon-wei replied:
@gianm Changed to regular HashMap

gianm commented on processing/src/test/java/io/druid/segment/IndexMergerTest.java:
```diff
@@ -741,22 +744,114 @@ public void testNonLexicographicDimOrderMerge() throws Exception
+    Assert.assertArrayEquals(new Object[]{3L}, boatList.get(2).getMetrics());
+  }
+
+  @Test
```
Could you add some tests involving null and empty string values?

One for merging one index without nulls or empty strings, and one index with nulls, and one index with empty strings, would be interesting if there isn't already a test like that.

jon-wei replied:
@gianm testMergeWithSupersetOrdering() in IndexMergerTest covered merging behavior with null values, I've added some additional logic to test that covers merging of empty strings

jon-wei commented:
@gianm Addressed your PR comments, I'm leaving the commit unsquashed for now for easier reference

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -555,19 +592,32 @@ private File makeIndexFiles(
       dimConversions.add(Maps.<String, IntBuffer>newHashMap());
     }
 
+    Map<String, Integer> nullStringIndexMap = Maps.<String, Integer>newHashMap();
```
I think this is always going to be 0 when a dimension has null values. If so then it could just be a Set, and the MMappedIndexRowIterable could use a constant array like `int[] NULL_VALUE = {0}`

gianm replied:
Or we could just *always* include empty string in the `dimensionValues`, and then we don't need this at all, since it'd be OK for MMappedIndexRowIterable to assume that null is 0.

gianm commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -608,6 +608,7 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
             continue;
           }
 
+          Integer emptyStrIdx = dictionary.indexOf("");
```
`indexOf` returns a primitive int, no need to use `Integer`.

gianm commented:
@jon-wei is it necessary to tweak IndexIO's "cardinality 0" check to avoid writing columns full of empty strings when there is no data for a dimension? Or is something else preventing that from happening?

gianm commented:
@jon-wei oh, is the "dimHasValues" check in IndexMerger the thing that makes that work? It looks like it is.

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -1183,12 +1241,20 @@ public Rowboat apply(@Nullable Rowboat input)
                 }
 
-                if (i >= dims.length || dims[i] == null) {
+                if (dims[i] == null) {
```
`dims[i]` will throw an exception if `i >= dims.length`

jon-wei replied:
@gianm my bad, I now moved that i >= dims.length check before the null check

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -562,13 +598,30 @@ private File makeIndexFiles(
+        }
+      }
+
+      // Ensure the empty str is always in the dictionary if column is not null across indexes
```
Comment here about *why* we need it would be good (at least mention that the MMappedIndexRowIterable depends on it for being able to equate missing dims with empty strings)

jon-wei replied:
@gianm Updated the comment to note that it's used by MMappedIndexRowIterable for merging missing dims with empty strings

jon-wei commented:
@gianm The "dimHasValues" check is responsible for avoiding writing a column if there was no data at all for a dimension across indexes

fjy commented:
I've reviewed IndexMerger and IndexIO as per request and I think it is cool :+1: on the changes there

gianm commented:
:+1: looks good to me

@jon-wei could you please rebase & squash & fix conflicts?
@himanshug @xvrl would either of you like to take another look as things have changed since your last review?

jon-wei commented:
@gianm Rebased and squashed

fjy commented:
@jon-wei Looks like some more conflicts

jon-wei commented:
merging, also updating IndexMergerV9 with similar logic

jon-wei commented:
Finished the merge and updated IndexMergerV9.

I've also updated IndexMergerV9Test such that the use of V9/legacy IndexMerger is controlled by a test parameter since the tests are identical to those in IndexMergerTest (now deprecated in the PR).

drcrallen commented:
@jon-wei is this a blocker for 0.9.0?

gianm commented:
@drcrallen I don't think it's a blocker but I think we should try to get it in.

gianm commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -644,7 +644,7 @@ private void mergeIndexesAndWriteColumns(
           continue;
         }
-        if (dims[i] == null || dims[i].length == 0) {
+        if (dims[i] == null || dims[i].length == 0 || dims[i][0] == 0) {
```
Is dim val 0 guaranteed to be a null/emptystr?

Even if it is, this last check should be `dims[i].length == 1 && dims[i][0] == 0` as the dim value could potentially be an empty str + something else.

gianm replied:
Looks like it is guaranteed due to the block below. but, the other check should still be modified, I think.

jon-wei replied:
@gianm added length == 1 check

gianm commented:
@jon-wei new changes look good other than the one comment I had

jon-wei commented:
rebased again

gianm commented:
:+1:

navis commented:
Sorry for late incursion but this makes invalid cardinality for segment meta query. If this could not be fixed we might rollback this.

jon-wei commented:
@navis Can you elaborate on the issue you are seeing?

navis commented:
with druid.sample.tsv,

before
```
io.druid.segment.IndexMerger - Starting dimension[market] with cardinality[3]
io.druid.segment.IndexMerger - Starting dimension[null_column] with cardinality[0]
io.druid.segment.IndexMerger - Starting dimension[partial_null_column] with cardinality[2]
io.druid.segment.IndexMerger - Starting dimension[placement] with cardinality[1]
io.druid.segment.IndexMerger - Starting dimension[placementish] with cardinality[9]
io.druid.segment.IndexMerger - Starting dimension[quality] with cardinality[9]
```

after
```
io.druid.segment.IndexMerger - Starting dimension[market] with cardinality[4]
io.druid.segment.IndexMerger - Starting dimension[null_column] with cardinality[0]
io.druid.segment.IndexMerger - Starting dimension[partial_null_column] with cardinality[2]
io.druid.segment.IndexMerger - Starting dimension[placement] with cardinality[2]
io.druid.segment.IndexMerger - Starting dimension[placementish] with cardinality[10]
io.druid.segment.IndexMerger - Starting dimension[quality] with cardinality[10]
```

navis commented:
I can see `null` is added to all dimensions with value. But it can make some problems in user side.
